\chapter{\label{ch:current-research}Current Research}

\minitoc

Research on automated splice site detection goes back to at least 1987. For
example, work \cite{shapiro1987rna} used the probability of occurrence of
particular nucleic bases at particular positions for splice site detection.

The hidden Markov model and the AdaBoost classifier to detect an intron splice
site were used in \cite{pashaei2016novel}. The hidden Markov model was also
used in \cite{goel2015improved} for sequence pre-processing and the support
vector machine for intron splice site detection. A multilayered recurrent
neural network applied on triplets from original sequence is described in
\cite{sarkar2019splice}. A convolutional neural network and its interpretation
is described in \cite{zuallaert2018splicerover}, which reports that the network
is most sensitive to nucleotides close to a splice site.

A paper \cite{hassanzadeh2016deeperbind} reports good performance of recurrent
convolutional neural networks, known as DeeperBind, in detecting the protein
binding motifs in DNA sequences. DeeperBind encodes DNA sequences with
one-hot-encoding. The network consists of one convolutional layer, followed by
two LSTM layers, and finally, the LSTM output is connected to two fully
connected layers with dropout. The convolutional layers are not followed by max
pooling layers common in CNN architectures. Strides of convolution is set to
one. The paper reports that the following LSTM layers are able to deal with
increased redundancy produced with convolutional layers with stride one and
without max pooling.

The paper \cite{lee2015dna} takes a different approach to RNA sequence
encoding. The authors report diminished generalization when learning
one-hot-encoded sequences and overcoming this issue by embedding each
nucleotide to a four-dimensional dense vector. The input is followed by two RNN
layers whose output is fed to a fully connected layer with dropout. The paper
compares the performances of ReLU, LSTM, and GRU recurrent units. It is
concluded that the addition of more RNN layers does not further improve the
performance and that LSTM units show the best performance.

Article \cite{naito2018human} describes the use of recurrent convolutional
neural networks for splice site detection in the human genome. Two
convolutional layers, each succeeded by a max-pooling layer, are followed by a
bidirectional LSTM recurrent layer. The output of the RNN layer is fed to two
fully connected layers with soft-max at the final output. Dropout is applied on
the outputs of all layers. One-hot-encoding is used on the inputs. The paper
reports improved performance of the network with the RNN layer as compared with
the same network without that layer.

Past research is mostly based on classical methods; the use of deep neural
networks for splice site detection is rare. Some methods show almost perfect
performance \cite{sarkar2019splice}. However, these methods are usually trained
and tested on a single or a few organisms---for example, Homo Sapiens---and
training-testing data is split on the level of individual samples as opposed to
organism or higher taxonomic ranks. The ability of these algorithms to
generalize to the DNA sequences of species not present in training data is
largely unreported.

It has been described that the splice site recognition process is complex and
tissue-specific in humans \cite{pineda2018most}. This complexity may limit the
maximum achievable accuracy of splice site detection algorithms that work with
DNA sequences only on out-of-sample organisms.

Convolutional neural networks are used in work \cite{senior2020improved} for
predicting the discrete probability distribution of distances and torsions of
amino acid residues based on pre-processed protein sequences. The neural
network is structurally similar to the networks used in image-recognition
tasks. The predicted distances were used as a seed for a gradient-descent
algorithm that predicts the three-dimensional structure of a protein by
minimizing the potential. The paper reports the state-of-the-art performance
and demonstrates the ability of deep neural networks to predict the physical
shapes of biological polymers.

The specifics of fungal DNA are exploited in the thesis. The ability of various
neural networks to generalize across various taxonomic ranks is studied. The
sheer amount of available annotated data enables the use of more complex neural
networks, which, in turn, can learn more subtle or complex features.
